File: PY/species_pointline_plot.py
#!/usr/bin/env python3
"""绘制两个物种丰度的散点回归图，并标注相关性与置信区间。"""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from scipy import stats

CURRENT_DIR = Path(__file__).resolve().parent
if str(CURRENT_DIR) not in sys.path:
    sys.path.insert(0, str(CURRENT_DIR))

from dataset import BioSmokeDataset
from split_group import prepare_crc_diff_groups

GROUP_COLORS = {
    'CRC_poordiff': '#D7263D',
    'CRC_welldiff': '#F18F01',
    'CTRL': '#2E86AB'
}
GROUP_ORDER = ["CTRL", "CRC_welldiff", "CRC_poordiff"]
GROUP_TITLE = {
    "CTRL": "CTRL",
    "CRC_welldiff": "CRC-Well",
    "CRC_poordiff": "CRC-Poor",
}

PLOT_STYLE = {
    "figure.facecolor": "white",
    "axes.facecolor": "white",
    "axes.edgecolor": "#202020",
    "axes.labelcolor": "#202020",
    "xtick.color": "#202020",
    "ytick.color": "#202020",
    "axes.grid": False,
    "font.size": 11,
    "axes.titlesize": 14,
    "axes.titleweight": "bold",
    "axes.labelsize": 12,
}


def _normalize_name(name: str) -> str:
    return re.sub(r"[^a-z0-9]+", "", str(name).lower())


def _italicize_species(name: str) -> str:
    text = str(name).strip().strip('"')
    text = re.sub(r"^(tax_|met_|kegg_)", "", text, flags=re.IGNORECASE)
    text = re.sub(r"^[a-z]__", "", text, flags=re.IGNORECASE)
    text = text.replace("_", " ")
    text = text.replace("\\", r"\\")
    text = text.replace(" ", r"\ ")
    return fr"$\it{{{text}}}$"


def _canonicalize_label(label: str) -> str:
    text = str(label).strip().strip('"')
    text = re.sub(r"^(tax_|met_|kegg_)", "", text, flags=re.IGNORECASE)
    text = re.sub(r"^[a-z]__", "", text, flags=re.IGNORECASE)
    return _normalize_name(text)


def _find_species_column(df: pd.DataFrame, target_name: str) -> str:
    # Align with R heatmap script, which uses tax_s__ species columns.
    species_cols = [col for col in df.columns if col.startswith("tax_s__")]
    if not species_cols:
        species_cols = [col for col in df.columns if col.startswith("tax_")]
    target_key = _canonicalize_label(target_name)

    exact_matches = [col for col in species_cols if _canonicalize_label(col) == target_key]
    if len(exact_matches) == 1:
        return exact_matches[0]
    if len(exact_matches) > 1:
        raise ValueError(f"物种 {target_name} 匹配到多个列: {exact_matches}")

    fallback_matches = [
        col for col in species_cols
        if target_key and target_key in _canonicalize_label(col)
    ]
    if len(fallback_matches) == 1:
        return fallback_matches[0]
    if len(fallback_matches) > 1:
        raise ValueError(f"物种 {target_name} 匹配到多个列: {fallback_matches}")

    available = ", ".join(species_cols[:20])
    raise ValueError(
        f"未找到物种 {target_name} 对应的列。可用物种列示例: {available}"
    )


def _spearman_stats_with_bootstrap_ci(
    x: pd.Series,
    y: pd.Series,
    confidence: float = 0.95,
    n_boot: int = 2000,
    random_state: int = 42,
) -> tuple[float, float, float, float]:
    rho, p_value = stats.spearmanr(x, y)
    n = len(x)

    if n <= 3 or np.isnan(rho):
        return float(rho), float(p_value), np.nan, np.nan

    rng = np.random.default_rng(random_state)
    x_arr = np.asarray(x)
    y_arr = np.asarray(y)
    boot_rhos = []

    for _ in range(n_boot):
        idx = rng.integers(0, n, size=n)
        boot_x = x_arr[idx]
        boot_y = y_arr[idx]
        boot_rho, _ = stats.spearmanr(boot_x, boot_y)
        if not np.isnan(boot_rho):
            boot_rhos.append(float(boot_rho))

    if len(boot_rhos) < 10:
        return float(rho), float(p_value), np.nan, np.nan

    alpha = 1.0 - confidence
    ci_low, ci_high = np.quantile(boot_rhos, [alpha / 2.0, 1.0 - alpha / 2.0])
    return float(rho), float(p_value), float(ci_low), float(ci_high)


def _prepare_group_data(
    df: pd.DataFrame,
    species_a: str,
    species_b: str,
    transform: str = "log1p",
) -> tuple[pd.DataFrame, str, str]:
    x_col = _find_species_column(df, species_a)
    y_col = _find_species_column(df, species_b)

    plot_df = df[[x_col, y_col, "group"]].copy()
    plot_df[x_col] = pd.to_numeric(plot_df[x_col], errors="coerce")
    plot_df[y_col] = pd.to_numeric(plot_df[y_col], errors="coerce")
    plot_df = plot_df.replace([np.inf, -np.inf], np.nan).dropna()

    plot_df = plot_df[plot_df["group"].isin(GROUP_ORDER)].copy()
    plot_df["group"] = pd.Categorical(plot_df["group"], categories=GROUP_ORDER, ordered=True)

    if transform == "log1p":
        plot_df[x_col] = np.log1p(plot_df[x_col].clip(lower=0))
        plot_df[y_col] = np.log1p(plot_df[y_col].clip(lower=0))
        x_label = f"{_italicize_species(species_a)} (log1p abundance)"
        y_label = f"{_italicize_species(species_b)} (log1p abundance)"
    elif transform == "none":
        x_label = _italicize_species(species_a)
        y_label = _italicize_species(species_b)
    else:
        raise ValueError("transform 仅支持 log1p 或 none")

    return plot_df, x_col, y_col, x_label, y_label


def _filter_display_outliers(
    df: pd.DataFrame,
    x_col: str,
    y_col: str,
    iqr_multiplier: float = 1.5,
    min_points: int = 12,
) -> tuple[pd.DataFrame, int]:
    """Filter extreme points for display only; keep full data for statistics."""
    if len(df) < 8:
        return df, 0

    q1_x, q3_x = df[x_col].quantile(0.25), df[x_col].quantile(0.75)
    q1_y, q3_y = df[y_col].quantile(0.25), df[y_col].quantile(0.75)
    iqr_x = q3_x - q1_x
    iqr_y = q3_y - q1_y

    if np.isclose(iqr_x, 0.0) and np.isclose(iqr_y, 0.0):
        return df, 0

    low_x = q1_x - iqr_multiplier * iqr_x
    high_x = q3_x + iqr_multiplier * iqr_x
    low_y = q1_y - iqr_multiplier * iqr_y
    high_y = q3_y + iqr_multiplier * iqr_y

    keep_mask = (
        (df[x_col] >= low_x)
        & (df[x_col] <= high_x)
        & (df[y_col] >= low_y)
        & (df[y_col] <= high_y)
    )
    filtered_df = df.loc[keep_mask].copy()
    removed = int((~keep_mask).sum())

    if len(filtered_df) < max(3, min_points):
        return df, 0
    return filtered_df, removed


def _plot_group_panel(
    ax: plt.Axes,
    plot_df: pd.DataFrame,
    x_col: str,
    y_col: str,
    group_name: str,
    x_label: str,
    y_label: str,
    show_xlabel: bool = True,
    show_ylabel: bool = True,
) -> None:
    group_df = plot_df[plot_df["group"] == group_name].copy()
    color = GROUP_COLORS[group_name]

    if len(group_df) < 3:
        ax.text(0.5, 0.5, "数据不足", transform=ax.transAxes, ha="center", va="center", fontsize=12)
        ax.set_axis_off()
        return

    rho, p_value, ci_low, ci_high = _spearman_stats_with_bootstrap_ci(group_df[x_col], group_df[y_col])

    # Use filtered points for visualization while keeping stats on full group_df.
    display_df, n_hidden = _filter_display_outliers(group_df, x_col, y_col)

    sns.regplot(
        data=display_df,
        x=x_col,
        y=y_col,
        ci=95,
        n_boot=1000,
        color=color,
        scatter_kws={
            "s": 44,
            "alpha": 0.82,
            "color": color,
            "edgecolor": "white",
            "linewidths": 0.45,
        },
        line_kws={"color": color, "lw": 2.6},
        ax=ax,
    )

    ax.set_title(f"{GROUP_TITLE[group_name]} (n={len(group_df)})", pad=12)
    ax.set_xlabel(x_label if show_xlabel else "")
    ax.set_ylabel(y_label if show_ylabel else "")

    annotation = (
        f"rho = {rho:.3f}\n"
        f"95% CI = [{ci_low:.3f}, {ci_high:.3f}]\n"
        f"p = {p_value:.2e}"
    )
    if n_hidden > 0:
        annotation += f"\nshown = {len(display_df)}/{len(group_df)}"
    ax.text(
        0.04,
        0.96,
        annotation,
        transform=ax.transAxes,
        ha="left",
        va="top",
        fontsize=10.5,
        bbox={"facecolor": "white", "alpha": 0.88, "edgecolor": color, "linewidth": 0.9, "pad": 5},
    )

    ax.grid(False)
    ax.set_facecolor("white")
    ax.set_frame_on(True)
    for spine in ax.spines.values():
        spine.set_visible(True)
        spine.set_linewidth(1.25)
        spine.set_color("#202020")
    ax.tick_params(axis="both", labelsize=10)


def _save_single_group_plot(
    plot_df: pd.DataFrame,
    x_col: str,
    y_col: str,
    species_a: str,
    species_b: str,
    group_name: str,
    output_dir: Path,
    x_label: str,
    y_label: str,
) -> Path:
    sns.set_theme(style="white", rc=PLOT_STYLE)
    fig, ax = plt.subplots(figsize=(6.9, 6.0), dpi=160)
    _plot_group_panel(ax, plot_df, x_col, y_col, group_name, x_label, y_label)
    ax.set_title(
        f"{GROUP_TITLE[group_name]}: {_italicize_species(species_a)} vs {_italicize_species(species_b)}",
        pad=14,
    )
    fig.tight_layout()

    safe_a = _normalize_name(species_a)
    safe_b = _normalize_name(species_b)
    plot_file = output_dir / f"{safe_a}_vs_{safe_b}_{group_name.lower()}_pointline.png"
    fig.savefig(plot_file, dpi=300, bbox_inches="tight")
    plt.close(fig)
    return plot_file


def build_species_pair_plot(
    species_a: str,
    species_b: str,
    output_dir: str,
    transform: str = "none",
) -> Path:
    ds = BioSmokeDataset()
    ds.preprocess_taxonomy_data(transform=True)
    ds.preprocess_metabolomics_data()
    df = ds.merge_to_dataframe()
    df = prepare_crc_diff_groups(df)

    plot_df, x_col, y_col, x_label, y_label = _prepare_group_data(df, species_a, species_b, transform)

    if len(plot_df) < 3:
        raise ValueError("可用于绘图的数据点少于 3 个，无法计算相关性")

    output_path = Path(output_dir)
    output_path.mkdir(parents=True, exist_ok=True)
    safe_a = _normalize_name(species_a)
    safe_b = _normalize_name(species_b)

    data_file = output_path / f"{safe_a}_vs_{safe_b}_pointline_data.csv"
    combined_file = output_path / f"{safe_a}_vs_{safe_b}_pointline_grouped.png"

    plot_df.to_csv(data_file, index=False)

    sns.set_theme(style="white", rc=PLOT_STYLE)
    fig, axes = plt.subplots(1, 3, figsize=(18.5, 6.0), dpi=160, sharex=True, sharey=True)
    if len(GROUP_ORDER) == 1:
        axes = [axes]

    for idx, group_name in enumerate(GROUP_ORDER):
        _plot_group_panel(
            axes[idx],
            plot_df,
            x_col,
            y_col,
            group_name,
            x_label,
            y_label,
            show_xlabel=True,
            show_ylabel=(idx == 0),
        )

    fig.suptitle(
        f"{_italicize_species(species_a)} vs {_italicize_species(species_b)}",
        fontsize=17,
        fontweight="bold",
        y=1.03,
    )
    fig.tight_layout()
    fig.savefig(combined_file, dpi=300, bbox_inches="tight")
    plt.close(fig)

    single_files = []
    for group_name in GROUP_ORDER:
        single_files.append(
            _save_single_group_plot(
                plot_df,
                x_col,
                y_col,
                species_a,
                species_b,
                group_name,
                output_path,
                x_label,
                y_label,
            )
        )

    print(f"已保存总览图像: {combined_file}")
    for plot_file in single_files:
        print(f"已保存分组图像: {plot_file}")
    print(f"已保存数据: {data_file}")
    return combined_file


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="绘制两个物种的点线相关图")
    parser.add_argument(
        "--species-a",
        default="Peptostreptococcus_stomatis",
        help="第一个物种名称",
    )
    parser.add_argument(
        "--species-b",
        default="Faecalibacterium_prausnitzii",
        help="第二个物种名称",
    )
    parser.add_argument(
        "--output-dir",
        default="results/species_pointline",
        help="输出目录",
    )
    parser.add_argument(
        "--transform",
        choices=["log1p", "none"],
        default="none",
        help="绘图前的数值变换方式",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    build_species_pair_plot(
        species_a=args.species_a,
        species_b=args.species_b,
        output_dir=args.output_dir,
        transform=args.transform,
    )


if __name__ == "__main__":
    main()

File: R/species_pointline_plot.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(ggplot2)
})

GROUP_COLORS <- c(
  CTRL = "#2E86AB",
  CRC_welldiff = "#F18F01",
  CRC_poordiff = "#D7263D"
)
GROUP_ORDER <- c("CTRL", "CRC_welldiff", "CRC_poordiff")
GROUP_TITLE <- c(
  CTRL = "CTRL",
  CRC_welldiff = "CRC-Well",
  CRC_poordiff = "CRC-Poor"
)

normalize_name <- function(x) {
  tolower(gsub("[^a-z0-9]+", "", as.character(x)))
}

canonicalize_label <- function(x) {
  text <- trimws(gsub('"', "", as.character(x)))
  text <- gsub("^(tax_|met_|kegg_)", "", text, ignore.case = TRUE)
  text <- gsub("^[a-z]__", "", text, ignore.case = TRUE)
  normalize_name(text)
}

pretty_species <- function(x) {
  text <- trimws(gsub('"', "", as.character(x)))
  text <- gsub("^(tax_|met_|kegg_)", "", text, ignore.case = TRUE)
  text <- gsub("^[a-z]__", "", text, ignore.case = TRUE)
  gsub("_", " ", text)
}

find_species_column <- function(df, target_name) {
  species_cols <- names(df)[grepl("^tax_", names(df))]
  target_key <- canonicalize_label(target_name)

  exact_matches <- species_cols[vapply(species_cols, function(col) {
    canonicalize_label(col) == target_key
  }, logical(1))]

  if (length(exact_matches) == 1) {
    return(exact_matches[1])
  }
  if (length(exact_matches) > 1) {
    stop(sprintf("Species %s matched multiple columns: %s",
                 target_name, paste(exact_matches, collapse = ", ")))
  }

  fallback_matches <- species_cols[vapply(species_cols, function(col) {
    nzchar(target_key) && grepl(target_key, canonicalize_label(col), fixed = TRUE)
  }, logical(1))]

  if (length(fallback_matches) == 1) {
    return(fallback_matches[1])
  }
  if (length(fallback_matches) > 1) {
    stop(sprintf("Species %s matched multiple columns: %s",
                 target_name, paste(fallback_matches, collapse = ", ")))
  }

  available <- paste(head(species_cols, 20), collapse = ", ")
  stop(sprintf("No column found for species %s. Example available columns: %s",
               target_name, available))
}

pearson_ci <- function(x, y, confidence = 0.95) {
  ct <- suppressWarnings(cor.test(x, y, method = "pearson"))
  r <- unname(ct$estimate)
  p_value <- ct$p.value
  n <- length(x)

  if (n <= 3 || isTRUE(all.equal(abs(r), 1))) {
    return(list(r = r, p = p_value, ci_low = NA_real_, ci_high = NA_real_))
  }

  clipped_r <- min(max(r, -0.999999), 0.999999)
  z <- atanh(clipped_r)
  se <- 1 / sqrt(n - 3)
  z_crit <- qnorm((1 + confidence) / 2)
  ci_low <- tanh(z - z_crit * se)
  ci_high <- tanh(z + z_crit * se)

  list(r = r, p = p_value, ci_low = ci_low, ci_high = ci_high)
}

prepare_crc_diff_groups <- function(df) {
  crc <- suppressWarnings(as.integer(df$crc_label))
  diff <- suppressWarnings(as.integer(df$differentiation))

  grp <- ifelse(is.na(crc), NA_character_, ifelse(
    crc == 1,
    ifelse(diff == 1, "CRC_poordiff", "CRC_welldiff"),
    "CTRL"
  ))

  df$group <- grp
  df <- df[!is.na(df$group), , drop = FALSE]
  df$group <- factor(df$group, levels = GROUP_ORDER, ordered = TRUE)
  df
}

prepare_plot_data <- function(df, species_a, species_b, transform = "log1p") {
  x_col <- find_species_column(df, species_a)
  y_col <- find_species_column(df, species_b)

  plot_df <- df[, c(x_col, y_col, "group"), drop = FALSE]
  names(plot_df) <- c("x", "y", "group")

  plot_df$x <- suppressWarnings(as.numeric(plot_df$x))
  plot_df$y <- suppressWarnings(as.numeric(plot_df$y))
  plot_df$x[is.infinite(plot_df$x)] <- NA_real_
  plot_df$y[is.infinite(plot_df$y)] <- NA_real_
  plot_df <- plot_df[complete.cases(plot_df), , drop = FALSE]

  plot_df <- plot_df[plot_df$group %in% GROUP_ORDER, , drop = FALSE]
  plot_df$group <- factor(as.character(plot_df$group), levels = GROUP_ORDER, ordered = TRUE)

  if (transform == "log1p") {
    plot_df$x <- log1p(pmax(plot_df$x, 0))
    plot_df$y <- log1p(pmax(plot_df$y, 0))
    x_label <- paste0(pretty_species(species_a), " (log1p abundance)")
    y_label <- paste0(pretty_species(species_b), " (log1p abundance)")
  } else if (transform == "none") {
    x_label <- pretty_species(species_a)
    y_label <- pretty_species(species_b)
  } else {
    stop("transform only supports: log1p or none")
  }

  list(plot_df = plot_df, x_col = x_col, y_col = y_col, x_label = x_label, y_label = y_label)
}

build_annotation <- function(df_group) {
  if (nrow(df_group) < 3) {
    return(list(text = "Insufficient data", n = nrow(df_group)))
  }

  stats_out <- pearson_ci(df_group$x, df_group$y, confidence = 0.95)
  ci_text <- if (is.na(stats_out$ci_low) || is.na(stats_out$ci_high)) {
    "95% CI = [NA, NA]"
  } else {
    sprintf("95%% CI = [%.3f, %.3f]", stats_out$ci_low, stats_out$ci_high)
  }

  txt <- paste(
    sprintf("r = %.3f", stats_out$r),
    ci_text,
    sprintf("p = %.2e", stats_out$p),
    sep = "\n"
  )

  list(text = txt, n = nrow(df_group))
}

make_single_plot <- function(group_df, group_name, x_label, y_label, title_text) {
  ann <- build_annotation(group_df)

  p <- ggplot(group_df, aes(x = x, y = y)) +
    geom_point(size = 2.2, alpha = 0.82, color = GROUP_COLORS[[group_name]]) +
    stat_smooth(method = "lm", se = TRUE, level = 0.95,
                color = GROUP_COLORS[[group_name]], linewidth = 0.9) +
    annotate("label", x = -Inf, y = Inf, hjust = -0.05, vjust = 1.05,
             label = ann$text, size = 3.2,
             label.size = 0.3,
             color = "black", fill = "white") +
    labs(
      title = sprintf("%s (n=%d)", GROUP_TITLE[[group_name]], ann$n),
      subtitle = title_text,
      x = x_label,
      y = y_label
    ) +
    theme_bw(base_size = 11) +
    theme(
      plot.title = element_text(face = "bold", size = 13),
      plot.subtitle = element_text(size = 10),
      axis.title = element_text(size = 11),
      panel.grid = element_blank(),
      panel.border = element_rect(color = "#202020", fill = NA, linewidth = 0.7)
    )

  p
}

build_species_pair_plot <- function(species_a,
                                    species_b,
                                    input_file = "dataset/merged_dataset_processed.csv",
                                    output_dir = "results/species_pointline",
                                    transform = "none") {
  df <- read.csv(input_file, check.names = FALSE, stringsAsFactors = FALSE)
  df <- prepare_crc_diff_groups(df)

  prepared <- prepare_plot_data(df, species_a, species_b, transform = transform)
  plot_df <- prepared$plot_df

  if (nrow(plot_df) < 3) {
    stop("Fewer than 3 points available for plotting.")
  }

  if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
  }

  safe_a <- normalize_name(species_a)
  safe_b <- normalize_name(species_b)

  data_file <- file.path(output_dir, sprintf("%s_vs_%s_pointline_data.csv", safe_a, safe_b))
  combined_file <- file.path(output_dir, sprintf("%s_vs_%s_pointline_grouped.png", safe_a, safe_b))

  write.csv(plot_df, data_file, row.names = FALSE)

  combined_plot <- ggplot(plot_df, aes(x = x, y = y, color = group)) +
    geom_point(size = 2.0, alpha = 0.82) +
    stat_smooth(method = "lm", se = TRUE, level = 0.95, linewidth = 0.85) +
    facet_wrap(~group, nrow = 1, scales = "fixed", labeller = as_labeller(GROUP_TITLE)) +
    scale_color_manual(values = GROUP_COLORS, drop = FALSE) +
    labs(
      title = sprintf("%s vs %s", pretty_species(species_a), pretty_species(species_b)),
      x = prepared$x_label,
      y = prepared$y_label,
      color = "Group"
    ) +
    theme_bw(base_size = 11) +
    theme(
      plot.title = element_text(face = "bold", size = 15, hjust = 0.5),
      strip.text = element_text(face = "bold", size = 11),
      panel.grid = element_blank(),
      panel.border = element_rect(color = "#202020", fill = NA, linewidth = 0.7),
      legend.position = "none"
    )

  ann_df <- do.call(rbind, lapply(GROUP_ORDER, function(g) {
    g_df <- plot_df[plot_df$group == g, , drop = FALSE]
    ann <- build_annotation(g_df)
    data.frame(group = g, label = ann$text, stringsAsFactors = FALSE)
  }))
  ann_df$group <- factor(ann_df$group, levels = GROUP_ORDER, ordered = TRUE)

  combined_plot <- combined_plot +
    geom_label(
      data = ann_df,
      aes(x = -Inf, y = Inf, label = label),
      inherit.aes = FALSE,
      hjust = -0.05,
      vjust = 1.05,
      size = 3.1,
      label.size = 0.3,
      fill = "white",
      color = "black"
    )

  ggsave(combined_file, combined_plot, width = 18.5, height = 6.0, dpi = 300)

  single_files <- character(0)
  for (group_name in GROUP_ORDER) {
    group_df <- plot_df[plot_df$group == group_name, , drop = FALSE]
    title_text <- sprintf("%s: %s vs %s",
                          GROUP_TITLE[[group_name]],
                          pretty_species(species_a),
                          pretty_species(species_b))
    p <- make_single_plot(group_df, group_name, prepared$x_label, prepared$y_label, title_text)

    out_file <- file.path(output_dir, sprintf("%s_vs_%s_%s_pointline.png",
                                              safe_a, safe_b, tolower(group_name)))
    ggsave(out_file, p, width = 6.9, height = 6.0, dpi = 300)
    single_files <- c(single_files, out_file)
  }

  message("Saved combined plot: ", combined_file)
  for (f in single_files) {
    message("Saved group plot: ", f)
  }
  message("Saved data: ", data_file)

  invisible(combined_file)
}

parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)

  defaults <- list(
    species_a = "Peptostreptococcus_stomatis",
    species_b = "Faecalibacterium_prausnitzii",
    input_file = "dataset/merged_dataset_processed.csv",
    output_dir = "results/R_plots/species_pointline",
    transform = "none"
  )

  if (length(args) == 0) {
    return(defaults)
  }

  i <- 1
  while (i <= length(args)) {
    key <- args[i]
    if (i == length(args)) {
      stop(sprintf("Missing value for argument: %s", key))
    }
    value <- args[i + 1]

    if (key == "--species-a") {
      defaults$species_a <- value
    } else if (key == "--species-b") {
      defaults$species_b <- value
    } else if (key == "--input-file") {
      defaults$input_file <- value
    } else if (key == "--output-dir") {
      defaults$output_dir <- value
    } else if (key == "--transform") {
      defaults$transform <- value
    } else {
      stop(sprintf("Unknown argument: %s", key))
    }
    i <- i + 2
  }

  if (!(defaults$transform %in% c("log1p", "none"))) {
    stop("--transform only supports: log1p or none")
  }

  defaults
}

main <- function() {
  args <- parse_args()
  build_species_pair_plot(
    species_a = args$species_a,
    species_b = args$species_b,
    input_file = args$input_file,
    output_dir = args$output_dir,
    transform = args$transform
  )
}

if (identical(environment(), globalenv())) {
  main()
}
